### PR TITLE
qemu: enable sandbox feature on qemu

### PR DESCRIPTION
**Pull request by LiangZhou-CTY**
Enable "-sandbox on" in qemu can introduce another protect layer
on the host, to make the secure container more secure.

The default option is disable because this feature may introduce some
performance cost, even though user can enable
/proc/sys/net/core/bpf_jit_enable to reduce the impact.

Fixes: #2266

Signed-off-by: Liang Zhou <zhoul110@chinatelecom.cn>

### CODE REVIEW

**Comment by jodh-intel** [comment]
Related: #1788.

/cc @ManaSugi.

File: src/runtime/Makefile [hunk 1]
```diff
@@ -164,6 +164,9 @@ DEFDISABLEGUESTSECCOMP := true
 #Default experimental features enabled
 DEFAULTEXPFEATURES := []
 
+#Default sandbox param
+DEFSANDBOXPARAM := [on,obsolete=deny,elevateprivileges=deny,spawn=deny,resourcecontrol=deny]
```
Comment (ManaSugi) [inline src/runtime/Makefile:168]: I think the square brackets `[]` are not needed.

```
DEFSANDBOXPARAM := on,obsolete=deny,elevateprivileges=deny,spawn=deny,resourcecontrol=deny
```

In addition, you need to add `DEFSANDBOXPARAM` to `USER_VAR` as follows in order to replace the variable in generated configuration files.

```
USER_VARS += DEFSANDBOXPARAM
```

Ref: https://github.com/kata-containers/kata-containers/blob/main/src/runtime/Makefile#L437

File: src/runtime/cli/config/configuration-qemu.toml.in [hunk 2]
```diff
@@ -314,10 +321,10 @@ valid_entropy_sources = @DEFVALIDENTROPYSOURCES@
 # the OCI spec passed to the runtime.
 #
 # You can create a rootfs with hooks by customizing the osbuilder scripts:
-# https://github.com/kata-containers/kata-containers/tree/main/tools/osbuilder
```
Comment (ManaSugi) [inline src/runtime/cli/config/configuration-qemu.toml.in:317]: This was fixed by [the commit](https://github.com/kata-containers/kata-containers/pull/2261).
Cloud you please put it back?

File: src/runtime/cli/config/configuration-qemu.toml.in [hunk 3]
```diff
@@ -56,6 +56,13 @@ firmware = "@FIRMWAREPATH@"
 # For example, `machine_accelerators = "nosmm,nosmbus,nosata,nopit,static-prt,nofw"`
 machine_accelerators="@MACHINEACCELERATORS@"
 
+# Qemu sandbox feature
+# comma-separated list of sandbox features to control the syscall access.
+# For example, `sandbox_on= "on,obsolete=deny,elevateprivileges=deny,spawn=deny,resourcecontrol=deny"`
+# Please note, enable this feature may lead some performance cost, you may enable
+# /proc/sys/net/core/bpf_jit_enable to reduce the impact.
+sandbox_on="@DEFSANDBOXPARAM@"
```
Comment (ManaSugi) [inline src/runtime/cli/config/configuration-qemu.toml.in:64]: This feature should be disabled by default as you mentioned, so you would need to comment out the line.

```
# sandbox_on="@DEFSANDBOXPARAM@"
```

**Review by ManaSugi — COMMENTED** [review]
Hi, @LiangZhou-CTY. I have a few comments.

File: src/runtime/cli/config/configuration-qemu.toml.in [hunk 4]
```diff
@@ -314,10 +321,10 @@ valid_entropy_sources = @DEFVALIDENTROPYSOURCES@
 # the OCI spec passed to the runtime.
 #
 # You can create a rootfs with hooks by customizing the osbuilder scripts:
-# https://github.com/kata-containers/kata-containers/tree/main/tools/osbuilder
+# https://github.com/kata-containers/osbuilder
 #
 # Hooks must be stored in a subdirectory of guest_hook_path according to their
-# hook type, i.e. "guest_hook_path/{prestart,poststart,poststop}".
```
Comment (ManaSugi) [inline src/runtime/cli/config/configuration-qemu.toml.in:320]: The same is true for this line.
Cloud you please put it back?

File: src/runtime/Makefile (hunk 1 above)
Comment (LiangZhou-CTY) [reply src/runtime/Makefile]: Thank you @ManaSugi , I've updated the code according to your comments.

File: src/runtime/cli/config/configuration-qemu.toml.in (hunk 2 above)
Comment (LiangZhou-CTY) [reply src/runtime/cli/config/configuration-qemu.toml.in]: Thank you @ManaSugi , I've updated the code according to your comments.

File: src/runtime/cli/config/configuration-qemu.toml.in (hunk 4 above)
Comment (LiangZhou-CTY) [reply src/runtime/cli/config/configuration-qemu.toml.in]: Thank you @ManaSugi , I've updated the code according to your comments.

File: src/runtime/vendor/github.com/kata-containers/govmm/qemu/qemu.go [hunk 5]
```diff
@@ -2440,6 +2440,9 @@ type Config struct {
 	// CPUModel is the CPU model to be used by qemu.
 	CPUModel string
 
+	// SandboxOn is the qemu function which enables the seccomp feature
+	SandboxOn string
+
```
Comment (fidencio) [inline src/runtime/vendor/github.com/kata-containers/govmm/qemu/qemu.go:2445]: Hey, this is not the way to update the vendored code.
You should submit a patch to https://github.com/kata-containers/govmm and then follow something like what has been done on 8385ff9554188d798fba07297240b7cb0feec251.

Comment (LiangZhou-CTY) [reply src/runtime/vendor/github.com/kata-containers/govmm/qemu/qemu.go:2445]: Thank you @fidencio for the comments. I raised https://github.com/kata-containers/govmm/pull/186 against govmm repo . 
After that PR is accepted I will be back here to continue .

Comment (fidencio) [reply src/runtime/vendor/github.com/kata-containers/govmm/qemu/qemu.go:2445]: Amazing, thanks @LiangZhou-CTY!
I started to review your PR there. :-)

**Comment by GabyCT** [comment]
@LiangZhou-CTY could you rebase your PR so conflicts are resolved, thanks

**Comment by LiangZhou-CTY** [comment]
> @LiangZhou-CTY could you rebase your PR so conflicts are resolved, thanks

Thank you @GabyCT  for your comments. Currently this PR depends on the another PR from govmm, so I will get back to this PR to handle the "sanity" work after govmm is merged.

**Pull request closed without merging**